src/libraries/Microsoft.PowerFx.Core/Texl/Builtins/Summarize.cs

```diff
@@ -0,0 +1,183 @@
+            : base("Summarize", TexlStrings.AboutSummarize, FunctionCategories.Table, DType.EmptyTable, 0, 2, int.MaxValue, DType.EmptyTable)
+        {
+            SignatureConstraint = new SignatureConstraint(omitStartIndex: 3, repeatSpan: 1, endNonRepeatCount: 1, repeatTopLength: 6);
+            ScopeInfo = new FunctionTableScopeInfo(this);
+        }
+
+        public override IEnumerable<TexlStrings.StringGetter[]> GetSignatures()
+        {
+            yield return new[] { TexlStrings.SummarizeArg1, TexlStrings.SummarizeArg2, TexlStrings.SummarizeArg3 };
+            yield return new[] { TexlStrings.SummarizeArg1, TexlStrings.SummarizeArg2, TexlStrings.SummarizeArg3, TexlStrings.SummarizeArg2 };
+            yield return new[] { TexlStrings.SummarizeArg1, TexlStrings.SummarizeArg2, TexlStrings.SummarizeArg2, TexlStrings.SummarizeArg3, TexlStrings.SummarizeArg3 };
+        }
+
+        // Produce a signature of the form:
+        // GroupBy(source, name, name, ..., name, name, groupName)
+        public override IEnumerable<TexlStrings.StringGetter[]> GetSignatures(int arity)
+        {
+            Contracts.Assert(arity >= 0);
+
+            if (arity <= 3)
+            {
+                return base.GetSignatures(arity);
+            }
+
+            // Limit the argCount avoiding potential OOM
+            int argCount = arity > SignatureConstraint.RepeatTopLength + SignatureConstraint.EndNonRepeatCount ? SignatureConstraint.RepeatTopLength + SignatureConstraint.EndNonRepeatCount : arity;
+            var args = new TexlStrings.StringGetter[argCount];
+
+            int lastIndex = argCount - 1;
+            args[0] = TexlStrings.SummarizeArg1;
+
+            var addSwitch = false;
+
+            for (int i = 1; i < lastIndex; i++)
+            {
+                if (addSwitch)
+                {
+                    args[i] = TexlStrings.SummarizeArg2;
+                }
+                else
+                {
+                    args[i] = TexlStrings.SummarizeArg3;
+                }
+
+                addSwitch = !addSwitch;
+            }
+
+            return new[] { args };
+        }
+
+        public override bool CheckTypes(CheckTypesContext context, TexlNode[] args, DType[] argTypes, IErrorContainer errors, out DType returnType, out Dictionary<TexlNode, DType> nodeToCoercedTypeMap)
+        {
+            Contracts.AssertValue(args);
+            Contracts.AssertAllValues(args);
+            Contracts.AssertValue(argTypes);
+            Contracts.Assert(args.Length == argTypes.Length);
+            Contracts.AssertValue(errors);
+            Contracts.Assert(MinArity <= args.Length && args.Length <= MaxArity);
+
+            bool isValid = base.CheckTypes(context, args, argTypes, errors, out returnType, out nodeToCoercedTypeMap);
+            Contracts.Assert(returnType.IsTable);
+
+            // Source table type
+            DType sourceType = argTypes[0];
+
+            if (!sourceType.IsTable)
```

Comment:
if it's not a table, should we just issue an error and stop?

Reply:
This can be removed. Other checks are catching this.